atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/lock/IsolatedProcessCorrectnessChecker.java

```diff
@@ -0,0 +1,138 @@
+                .build();
+    }
+
+    private static class Visitor implements EventVisitor {
+        private final Map<Pair<Integer, String>, Boolean> refreshAllowed = new HashMap<>();
+        private final Map<Pair<Integer, String>, OkEvent> lastEvent = new HashMap<>();
+        private final List<Event> errors = new ArrayList<>();
+        private final Map<Integer, String> resourceName = new HashMap<>();
+
+        @Override
+        public void visit(InfoEvent event) {
+        }
+
+        @Override
+        public void visit(InvokeEvent event) {
+            resourceName.put(event.process(), event.value());
+        }
+
+        /**
+         * LOCK:
+         *  - SUCCESS: held the lock at some point between request and reply.
+         *  - FAILURE: the lock was held by someone else at some point between request and reply.
+         *
+         * UNLOCK:
+         *  - SUCCESS: was holding the lock, now does not hold it.
+         *  - FAILURE: was not holding the lock.
+         *
+         * REFRESH:
+         *  - SUCCESS: was holding the lock, still possibly does.
+         *  - FAILURE: was not holding the lock.
+         */
+        @Override
+        public void visit(OkEvent event) {
+            int currentProcess = event.process();
+            String lockName = resourceName.get(currentProcess);
+            Pair processLock = new Pair(currentProcess, lockName);
+
+            if (!refreshAllowed.containsKey(processLock)) {
+                refreshAllowed.put(processLock, false);
+                lastEvent.put(processLock, event);
+            }
+
+            switch (event.function()) {
+                case RequestType.LOCK:
+                    if (event.isFailure()) {
+                        refreshAllowed.put(processLock, false);
+                    }
+                    if (event.isSuccessful()) {
+                        refreshAllowed.put(processLock, true);
+                    }
+                    break;
+                case RequestType.REFRESH:
+                    if (event.isFailure()) {
+                        refreshAllowed.put(processLock, false);
+                    }
+                    if (event.isSuccessful()) {
+                        if (!refreshAllowed.get(processLock)) {
+                            errors.add(lastEvent.get(processLock));
+                            errors.add(event);
+                            refreshAllowed.put(processLock, true);
+                        }
+                    }
+                    break;
+                case RequestType.UNLOCK:
+                    if (event.isSuccessful()) {
+                        if (!refreshAllowed.get(processLock)) {
+                            errors.add(lastEvent.get(processLock));
+                            errors.add(event);
+                        }
+                    }
+                    refreshAllowed.put(processLock, false);
+                    break;
+                default: break;
```

Comment:
I'd say yes, and in fact we should throw `IllegalStateException` or something else in this case. This will help if we say try to feed a timestamp history into the lock checker - we'll currently get valid since there are no meaningful events 😬